feat: Introduce task-specific permissions

ce-dmelnych opened the pull request:
## Description

This PR introduces task-specific permissions, enabling fine-grained access control for user task operations. It adds two new permission types (`CLAIM` and `COMPLETE`) and introduces the `USER_TASK` resource type with support for `READ`, `UPDATE`, `CLAIM`, and `COMPLETE` permissions. Default roles are updated automatically: `admin` role receives full permissions while `readonly-admin` role receives READ permission only.

**What this PR solves:**

This PR lays the foundation for managing granular permissions for user tasks, enabling to differentiate between general task updates and specific worker actions like task claiming and completion. It establishes the groundwork for implementing task-specific authorization checks for task operations.

## Related issues

closes #40015

ce-dmelnych committed:
feat: add CLAIM and COMPLETE permission types

Introduce new permission types that will be used for granular user task access control:
- CLAIM: Allows users to claim tasks(self assignment)
- COMPLETE: Allows users to complete tasks

These permissions will enable fine-grained control over user task operations
beyond basic READ and UPDATE permissions.

ce-dmelnych committed:
feat: add USER_TASK resource with permission bindings

Introduce USER_TASK as a new authorization resource type with support
for READ, UPDATE, CLAIM, and COMPLETE permissions.

ce-dmelnych committed:
test: verify USER_TASK permissions in default roles

Update IdentitySetupInitializeDefaultsTest to assert that:
- admin role receives READ, UPDATE, CLAIM, COMPLETE permissions for USER_TASK
- readonly-admin role receives READ permission for USER_TASK

Ensures default roles are correctly initialized with USER_TASK permissions.

ce-dmelnych commented on security/security-protocol/src/main/java/io/camunda/zeebe/protocol/record/value/AuthorizationResourceType.java:
```diff
@@ -60,7 +60,9 @@ public enum AuthorizationResourceType {
   UNSPECIFIED(),
-  USER(PermissionType.CREATE, PermissionType.READ, PermissionType.UPDATE, PermissionType.DELETE);
+  USER(PermissionType.CREATE, PermissionType.READ, PermissionType.UPDATE, PermissionType.DELETE),
+  USER_TASK(
```
Good question! The idea is to have task-scoped permissions that will coexist with `PROCESS_DEFINITION[READ_USER_TASK, UPDATE_USER_TASK]`.

The separation serves several purposes. `PROCESS_DEFINITION` permissions will be used by Team managers(supervisors), while `USER_TASK` permissions will be task-specific with lower precedence than `PROCESS_DEFINITION`, allowing for a clear override mechanism.

Having a dedicated `USER_TASK` resource type will make it possible to create authorizations for specific tasks (by `userTaskKey`), which isn't feasible with process definition-level permissions. This also allows us to track them separately and handle them differently in the UI (see the [design doc](https://docs.google.com/document/d/1HcGgSi3v1wDci2YSJMi8SSkENjtHSjMLzd-1Ei3016k/edit?tab=t.0#bookmark=id.5x6w3ktyyfvr) for reference).

So both resource types will work together to provide flexibility at different authorization levels.

ce-dmelnych replied:
As I'm currently aware, the UI wouldn't explicitly show that `PROCESS_DEFINITION` has precedence above other permissions, at least not for the MVP phase. However, we can definitely address this through documentation to avoid any confusion.

We could add a dedicated section in the [documentation](https://docs.camunda.io/docs/next/components/concepts/access-control/authorizations/#available-resources) with a clear description of the `USER_TASK` resource, its permissions, and how their precedence works in relation to `PROCESS_DEFINITION[READ_USER_TASK, UPDATE_USER_TASK]`. This way users would know what to expect when working with both resource types, and it shouldn't come as a surprise to them or cause unexpected behaviors for our support and medic teams.

@OmranAbazid what's your opinion about this?

tmetzke reviewed (commented):
👍 API looks good to me
💭 While we're touching those enums, we might as well sort them so it's easier to find what you're looking for.

korthout reviewed (approved):
Nice work @ce-dmelnych 👏 

❓ The adjusted tests show that the admin and read-only admin roles have the correct initial setup. Should we also add a test case that shows that these new permissions can be set up for other entities (roles, users, etc)?

👍 All other questions I had have been covered already by others and answered sufficiently by you.

🔧 I agree with tmetzke that we can sort the enums, but it's not a must.

ce-dmelnych committed:
refactor: sort ResourceType enum values alphabetically

Sort ResourceType enum values in alphabetical order for better
readability and consistency with documentation structure.

ce-dmelnych commented on zeebe/gateway-protocol/src/main/proto/rest-api.yaml:
```diff
@@ -8512,6 +8514,7 @@ components:
         - DECISION_DEFINITION
         - GROUP
         - USER
+        - USER_TASK
```
Yeah, at first glance it looked like there was some structure it their ordering, like `DELETE_*` permissions placed at the bottom, `READ_*` before `UPDATE_*`, but that might’ve just happened by accident 😄

As for `CLAIM` and `COMPLETE` we can consider them as a sort of `UPDATE`-like permissions, so, I've placed them close to them.

korthout reviewed (approved):
Thanks @ce-dmelnych 🚀 

LGTM 👍

korthout commented on clients/java/src/main/java/io/camunda/client/api/search/enums/PermissionType.java:
Counter question: Is the grouping useful in code? I see that it makes sense in the docs (per authorization the permissions are not alphabetically sorted by semantically, i.e. read before delete).

ce-dmelnych committed:
refactor: sort permission types alphabetically

Sort permission types alphabetically to improve code readability and maintainability.

Ben-Sheppard reviewed (approved):
🚀 LGTM
